File: Neetcode/2DDP/burstballoons.py
# https://neetcode.io/problems/burst-balloons
# https://leetcode.com/problems/burst-balloons

from typing import List


class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + [n for n in nums if n != 0] + [1]
        n = len(nums)
        # dp[l][r] is the maximum coins obtained from bursting all the balloons between nums[l] and nums[r], inclusive
        dp = [[0] * n for _ in range(n)]

        for off in range(1, n - 1):  # offset of the subproblem
            for l in range(1, n - off):  # left boundary of the subproblem
                r = l + off - 1  # right boundary of the subproblem
                for i in range(
                    l, r + 1
                ):  # i is the last balloon to burst in the current subproblem
                    dp[l][r] = max(
                        dp[l][r],
                        dp[l][i - 1]
                        + nums[l - 1] * nums[i] * nums[r + 1]
                        + dp[i + 1][r],
                    )
        return dp[1][n - 2]


class SlowSolution:
    def maxCoins(self, nums: List[int]) -> int:
        n = len(nums)
        nums = [1] + nums + [1]
        track = {}

        def dfs(l, r):
            if (l, r) in track:
                return track[(l, r)]
            if l == r:
                track[(l, r)] = nums[l - 1] * nums[l] * nums[r + 1]
                return track[(l, r)]

            ret = 0
            for i in range(l, r + 1):
                ret = max(
                    ret,
                    (
                        nums[l - 1] * nums[i] * nums[r + 1]
                        + dfs(l, i - 1)
                        + dfs(i + 1, r)
                    ),
                )

            track[(l, r)] = ret
            return ret

        return dfs(1, n)

File: Neetcode/2DDP/burstballoons.cpp
// https://neetcode.io/problems/burst-balloons
// https://leetcode.com/problems/burst-balloons

#include <vector>
using namespace std;

class Solution {
 public:
  int maxCoins(vector<int>& nums) {
    // Add [1] to the beginning and end of the array
    nums.insert(nums.begin(), 1);
    nums.push_back(1);
    int n = nums.size();

    // dp[l][r] = max coins obtained by bursting balloons from l to r
    vector<vector<int>> dp(n, vector<int>(n, 0));
    int r, mult;

    // off is the distance between l and r, we loop through all possible lengths
    //  of subarrays in nums, this allows us to do tabloid dp
    for (int off = 1; off <= n - 2; off++) {
      for (int l = 1; l < n - off; l++) {
        r = l + off - 1;
        mult = nums[l - 1] * nums[r + 1];
        for (int i = l; i <= r; i++) {
          dp[l][r] =
              max(dp[l][r], dp[l][i - 1] + nums[i] * mult + dp[i + 1][r]);
        }
      }
    }

    // Ans is the maximum coins obtained by bursting balloons from 1 to n - 2,
    //  ignoring added 1s
    return dp[1][n - 2];
  }
};
